zk/tx/tx.go

```diff
@@ -77,20 +77,22 @@ func DecodeBatchL2Blocks(txsData []byte, forkID uint64) ([]DecodedBatchL2Data, e
 
 		// if num is 11 then we are trying to parse a `changeL2Block` transaction
 		if num == changeL2BlockTxType {
-			currentDelta = binary.BigEndian.Uint32(txsData[pos+1 : pos+5])
-			currentL1InfoTreeIndex = binary.BigEndian.Uint32(txsData[pos+5 : pos+9])
-
-			pos += changeL2BlockLength
+			hasStarted := pos > 0
 
 			// check if we're at the end of the line with an empty block and add it or if the next byte is a change block transaction
 			// we want to capture the data immediately and add it to the result
-			if pos == txDataLength || txsData[pos] == changeL2BlockTxType {
+			if hasStarted || txsData[pos+1] == changeL2BlockTxType {
```

Comment:
this change is unexpected, why is that? maybe some comment is worthy

Reply:
This simplifies the logic, we now capture data any time we find a changeL2Block transaction unless it's the very first one at the start of the data.  We now also capture the data outside of the loop to mop up the trailing data.  This allows us to cleanly handle scenarios with no changeL2Block data (as from the datastream) and where they are present (l1 data) in the same path.  There are a whole suite of tests over this now to handle empty blocks at the start, end, in the middle of the data etc. etc.

Reply:
I have added a more substantial comment to this code now as well